src/rgw/rgw_file.cc

```diff
@@ -926,7 +926,212 @@ namespace rgw {
+	lsubdout(get_context(), rgw, 15)
+	  << "\tkey: " << attr.first << dendl;
+      }
+    }
+
+    rc = rgwlib.get_fe()->execute_req(&req);
+    rc2 = req.get_ret();
+    rc3 = ((rc == 0) && (rc2 == 0)) ? 0 : -EIO;
+
+    /* call back w/xattr data--signal eof? */
+    if (rc3 == 0) {
+      const auto& attrs = req.get_attrs();
+      for (const auto& attr : attrs) {
+
+	if (!attr.second.has_value())
+	  continue;
+
+	const auto& k = attr.first;
+	const auto& v = attr.second.value();
+
+	std::string_view svk = unprefix_xattr_keystr(k);
+	rgw_xattrstr xattr_k = { uint32_t(svk.length()),
+				 const_cast<char*>(svk.data()) };
+	rgw_xattrstr xattr_v = { uint32_t(v.length()),
+				 const_cast<char*>(
+				   const_cast<buffer::list&>(v).c_str()) };
+	rgw_xattr xattr = { xattr_k, xattr_v };
+	rgw_xattrlist xattrlist = { &xattr, 1 };
+
+	cb(&xattrlist, cb_arg, RGW_GETXATTR_FLAG_NONE);
+      }
+    }
+
+    return rc3;
+  } /* RGWLibFS::getxattrs */
+
+  int RGWLibFS::lsxattrs(RGWFileHandle* rgw_fh, rgw_xattrstr *start_at,
+			 rgw_xattrstr *filter_prefix /* ignored */,
+			 rgw_getxattr_cb cb, void *cb_arg,
+			 uint32_t flags)
+  {
+    /* cannot store on fs_root, should not on buckets? */
+    if ((rgw_fh->is_bucket()) ||
+	(rgw_fh->is_root()))  {
+      return -EINVAL;
+    }
+
+    int rc, rc2, rc3;
+    string obj_name{rgw_fh->relative_object_name2()};
+
+    /* XXX prefixing */
+    string marker{start_at->val, start_at->len};
+
+    RGWGetAttrsRequest req(cct, rgwlib.get_store()->get_user(user.user_id),
+			   rgw_fh->bucket_name(), obj_name,
+			   marker, true /* keys_only */);
+
+    rc = rgwlib.get_fe()->execute_req(&req);
+    rc2 = req.get_ret();
+    rc3 = ((rc == 0) && (rc2 == 0)) ? 0 : -EIO;
+
+    /* call back w/xattr data--signal eof? */
+    if (rc3 == 0) {
+      const auto& keys = req.get_attrs();
+      for (const auto& k : keys) {
+	std::string_view svk = unprefix_xattr_keystr(k.first);
+	rgw_xattrstr xattr_k = { uint32_t(svk.length()),
+				 const_cast<char*>(svk.data()) };
+	rgw_xattrstr xattr_v = { 0, nullptr };
+	rgw_xattr xattr = { xattr_k, xattr_v };
+	rgw_xattrlist xattrlist = { &xattr, 1 };
+
+	cb(&xattrlist, cb_arg, RGW_GETXATTR_FLAG_NONE);
+      }
+    }
+
+    return rc3;
+  } /* RGWLibFS::lsxattrs */
+
+  int RGWLibFS::setxattrs(RGWFileHandle* rgw_fh, rgw_xattrlist *attrs,
+			  uint32_t flags)
+  {
+    /* cannot store on fs_root, should not on buckets? */
+    if ((rgw_fh->is_bucket()) ||
+	(rgw_fh->is_root()))  {
+      return -EINVAL;
+    }
+
+    int rc, rc2;
+    string obj_name{rgw_fh->relative_object_name2()};
+
+    RGWSetAttrsRequest req(cct, rgwlib.get_store()->get_user(user.user_id),
+			   rgw_fh->bucket_name(), obj_name);
+
+    for (uint32_t ix = 0; ix < attrs->xattr_cnt; ++ix) {
+      auto& xattr = attrs->xattrs[ix];
+      buffer::list attr_bl;
+      string k = prefix_xattr_keystr(xattr.key);
+      /* don't allow storing at RGW_ATTR_META_PREFIX */
+      if (! (k.length() > 0)) {
```

Comment:
Should this check xattr.key instead of k?  k should always have RGW_ATTR_META_PREFIX in it, and therefore be > 0?

Reply:
oops, thanks